zuul-core/src/jmh/java/com/netflix/zuul/message/HeadersBenchmark.java

```diff
@@ -0,0 +1,167 @@
+
+    @State(Scope.Thread)
+    public static class AddHeaders {
+        @Param({"0", "1", "5", "10", "30"})
+        public int count;
+
+        @Param({"10"})
+        public int nameLength;
+
+        private String[] stringNames;
+        private HeaderName[] names;
+        private String[] values;
+
+        @Setup
+        public void setUp() {
+            stringNames = new String[count];
+            names = new HeaderName[stringNames.length];
+            values = new String[stringNames.length];
+            for (int i = 0; i < stringNames.length; i++) {
+                UUID uuid = new UUID(ThreadLocalRandom.current().nextLong(), ThreadLocalRandom.current().nextLong());
+                String name = uuid.toString();
+                assert name.length() >= nameLength;
+                name = name.substring(0, nameLength);
+                names[i] = new HeaderName(name);
+                stringNames[i] = name;
+                values[i] = name;
+            }
+        }
+
+        @Benchmark
+        @BenchmarkMode(Mode.AverageTime)
+        @OutputTimeUnit(TimeUnit.NANOSECONDS)
+        public Headers addHeaders_string() {
+            Headers headers = new Headers();
+            for (int i = 0; i < count; i++) {
+                headers.add(stringNames[i], values[i]);
+            }
+            return headers;
+        }
+
+        @Benchmark
+        @BenchmarkMode(Mode.AverageTime)
+        @OutputTimeUnit(TimeUnit.NANOSECONDS)
+        public Headers addHeaders_headerName() {
+            Headers headers = new Headers();
+            for (int i = 0; i < count; i++) {
+                headers.add(names[i], values[i]);
+            }
+            return headers;
+        }
+    }
+
+
+    @State(Scope.Thread)
+    public static class GetSetHeaders {
+        @Param({"1", "5", "10", "30"})
+        public int count;
+
+        @Param({"10"})
+        public int nameLength;
+
+        private String[] stringNames;
+        private HeaderName[] names;
+        private String[] values;
+        Headers headers;
+
+        @Setup
+        public void setUp() {
+            headers = new Headers();
+            stringNames = new String[count];
+            names = new HeaderName[stringNames.length];
+            values = new String[stringNames.length];
+            for (int i = 0; i < stringNames.length; i++) {
+                UUID uuid = new UUID(ThreadLocalRandom.current().nextLong(), ThreadLocalRandom.current().nextLong());
+                String name = uuid.toString();
+                assert name.length() >= nameLength;
+                name = name.substring(0, nameLength);
+                names[i] = new HeaderName(name);
+                stringNames[i] = name;
+                values[i] = name;
+                headers.add(names[i], values[i]);
+            }
+        }
+
+        @Benchmark
+        @BenchmarkMode(Mode.AverageTime)
+        @OutputTimeUnit(TimeUnit.NANOSECONDS)
+        public void setHeader_first() {
```

Comment:
Not sure I'm clear about the hypothesis around benchmarking `setHeader_` and `getHeader`?

Reply:
See the new PR I sent. :)